IPNI_HTTP_PROVIDER.md

```diff
@@ -0,0 +1,310 @@
+* `application/cbor` is considered to be the same as `application/vnd.ipld.dag-cbor`.
+
+The common data types, such as `CID` and `Multihash`, are encoded in their
+standard [multibase format](https://github.com/multiformats/multibase) when referenced in URLs.
+
+### Versioning
+
+The IPNI Provider HTTP API adopts a versioning strategy that utilizes the path prefix to indicate major releases. As a
+best practice, future iterations that introduce breaking changes should increment the version number in the API path.
+This approach ensures compatibility and allows clients to adapt to new versions while maintaining backward compatibility
+with existing implementations. By explicitly incorporating versioning in the API path, it becomes easier to manage and
+track changes over time, facilitating smooth transitions and providing a clear indication of API evolution.
+
+
+### Namespacing
+
+This specification introduces the `/ipni/` URL prefix to namespace the functionality associated with IPNI HTTP providers. The presence of this namespace allows IPNI specifications to evolve without conflicting with other URL patterns that an HTTP server may support. Therefore, this specification reserves the right for any sub-tree that appears after this prefix.
+
+#### Implicit Namespace Assumption
+
+Indexer nodes implicitly assume the `/ipni/` URL prefix. When receiving an announcement from providers, an indexer will attempt to fetch advertisements by joining the announced URL with the paths documented in this specification. Implementers have the freedom to include additional paths to the announced URL if desired.
+
+For example, if the announced URL is `https://ipni-provider.example`, it will be accessed at `https://ipni-provider.example/ipni/v1/ad/head` to fetch the head advertisement. Similarly, if the announced URL is `https://ipni-provider.example/my/prefix`, it will be accessed at `https://ipni-provider.example/my/prefix/ipni/v1/ad/head`.
+
+#### Custom Paths in Multiaddr
+
+IPNI implementers have the option to announce HTTP publisher URLs in `Multiaddr` format. However, this format does not officially support paths. To accommodate custom paths in the announced URLs, implementers can utilize the `httpath` parameter. The handling of `httpath` is similar to that of vanilla HTTP URLs with an additional specified path.
+
+For example, if the announced multiaddr is `/dns/ipni-provider.example/https`, it will be accessed at `https://ipni-provider.example/ipni/v1/ad/head` to fetch the head advertisement. On the other hand, if the announced multiaddr is `/dns/ipni-provider.example/https/httpath/%2Fmy%2Fprefix`, it will be accessed at `https://ipni-provider.example/my/prefix/ipni/v1/ad/head`.
+
+By supporting custom paths through `httpath`, IPNI implementers can enhance the flexibility of their HTTP publisher URLs while maintaining compatibility with the IPNI HTTP Provider API.
+
+## Specification
+
+### GET `/ipni/v1/ad/head`
```

Comment:
I assume this either has a limited TTL where a stale response can be cached for some time,  or should not be cached at all, so the client gets up-to-date response.

Either way, you should require implementer to always return explicit `Cache-Control`:

- To disable HTTP caching and always fetch the latest response:
  ```
  Cache-Control: no-cache, no-store, must-revalidate
  ```
- To allow accepting a cached response and revalidating it in the background, you can use the following Cache-Control header:
   ```
   Cache-Control: max-age=60, stale-while-revalidate=120
   ```
   if the cached response is stale (older than 60 seconds), the client can still serve it while simultaneously sending a request to the server to get updated one -- this removes spikes in latency when Cloudfront  experiences cache miss due to cache age